### PR TITLE
Benchmark/code generation from profiles

### PR DESCRIPTION
**Pull request by dmos62**
WIP. At the time of writing a code render request and response look like this:

```
> curl 'http://localhost:4040/render?from=now-1h&until=now&query=pyroscope.server.alloc_objects&format=code&lang=go'
package main
import "time"

var fullCycleInMillis = 100
var cpuUtilizationTarget = 0.800000
var durationToWork = time.Duration(int(cpuUtilizationTarget * float64(fullCycleInMillis))) * time.Millisecond
var durationToSleep = time.Duration(int((1 - cpuUtilizationTarget) * float64(fullCycleInMillis))) * time.Millisecond
var durationToSleepChannel = make(chan time.Duration, 1)
func switchToSleep() {
	durationToSleepChannel <- durationToSleep
	time.AfterFunc(durationToWork, switchToSleep)
}
var shouldDoSetup = true
func setupFunction() {
	if shouldDoSetup {
		time.AfterFunc(durationToWork, switchToSleep)
		shouldDoSetup = false
	}
}
		
func f1e02707a0b8ef39908b2f29b9814378d67acfe522b7e968640f860a5() {
	f35b4d1b3e72cbe285ada58ac45d402fc1d6ba8b29429b57ae8ce2e03()
	fa71d2e68f2a95e01175045d822c7fb488b88f9c304ce689bf43c7375()
}
func f35b4d1b3e72cbe285ada58ac45d402fc1d6ba8b29429b57ae8ce2e03() {
	fe97a65f831efed5cb2b676120f3de95ffc2e0e69e9518a7004e71c39()
}
func fe97a65f831efed5cb2b676120f3de95ffc2e0e69e9518a7004e71c39() {
	f64a631cc6cfbfd661de434f61ebcceeb8fe7de37dd31295ae191d745()
}
func f64a631cc6cfbfd661de434f61ebcceeb8fe7de37dd31295ae191d745() {
	f379bda7c80b12f29e17ec31e23581588594c7ca34d88233894010e60()
}
func f379bda7c80b12f29e17ec31e23581588594c7ca34d88233894010e60() {
	f9a1d18f2d39012ac1bca9908775d528c06234a405ee223147d60a3ae()
}
func f9a1d18f2d39012ac1bca9908775d528c06234a405ee223147d60a3ae() {
	fc4154700455823487946fa06e018e01351ab5f247e6b49c6ff9bcf0f()
}
func fc4154700455823487946fa06e018e01351ab5f247e6b49c6ff9bcf0f() {
	fd2bd46c273972891535f0713e502cd7ff73e5d46581415ba8ae1c7fd()
	ff1e9eba88f19b4842d0869e5e7abaa11843d8a4be77b489ca6f9eaee()
}
func fd2bd46c273972891535f0713e502cd7ff73e5d46581415ba8ae1c7fd() {
	fadffd9b02047dae722b9cde4d603c5af5de8a380b769edda7202827f()
}
func fadffd9b02047dae722b9cde4d603c5af5de8a380b769edda7202827f() {
	f4f93af98460f1dc6116568e635352433d0e4bc1005b5c050378a4533()
}
func f4f93af98460f1dc6116568e635352433d0e4bc1005b5c050378a4533() {
	f96d72aed3a8c2596da0d55be820693397457c8442019155330aaf4a3()
}
func f96d72aed3a8c2596da0d55be820693397457c8442019155330aaf4a3() {
	f134c8bdd48c0c5911ee2d86e1188cd8aed6c0758736f019e0adc3c29()
}
func f134c8bdd48c0c5911ee2d86e1188cd8aed6c0758736f019e0adc3c29() {
	f3c2d1d3e40ac0af275f92d48934a50c0efc986697ab605f102b3fc3e()
}
func f3c2d1d3e40ac0af275f92d48934a50c0efc986697ab605f102b3fc3e() {
	for i := 0; i < 387; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
}
func ff1e9eba88f19b4842d0869e5e7abaa11843d8a4be77b489ca6f9eaee() {
	f9beabf9c2ab6588456b488647af7438f7d50fae5863917dd240da931()
}
func f9beabf9c2ab6588456b488647af7438f7d50fae5863917dd240da931() {
	f22fed946ed0953ad00a7b1c22992f90664b1a59bbb247e75504416be()
}
func f22fed946ed0953ad00a7b1c22992f90664b1a59bbb247e75504416be() {
	f02e4f73dbf14b8e2c26dc73fdd7993ff8048a7d55e79886fd8d8f4d1()
	ffb90608b44eeb3d6e6440c4e314ac694c6ed2aadd92d4b169f4ae25c()
}
func f02e4f73dbf14b8e2c26dc73fdd7993ff8048a7d55e79886fd8d8f4d1() {
	fa81652b9f605dc5546bad23a74f7ec7bbe73e2af65419e79d82bf997()
}
func fa81652b9f605dc5546bad23a74f7ec7bbe73e2af65419e79d82bf997() {
	fea78efc85ad98d388fe3b8f0e3868e7de0939c3773566dfa5705557d()
}
func fea78efc85ad98d388fe3b8f0e3868e7de0939c3773566dfa5705557d() {
	ff1f28b2f23c435c6614047ccd6ee9c1e4f96c04ba1cd59b4c1f34009()
}
func ff1f28b2f23c435c6614047ccd6ee9c1e4f96c04ba1cd59b4c1f34009() {
	fde9829f49e3b2681abce0bd9f488370321b8b4f23d7076c2ae99151c()
}
func fde9829f49e3b2681abce0bd9f488370321b8b4f23d7076c2ae99151c() {
	fbbab1391f758f111eb7d66ab69983f9ec0a119fb0ffb63adf6672f63()
}
func fbbab1391f758f111eb7d66ab69983f9ec0a119fb0ffb63adf6672f63() {
	f57b7ab9bf422171778cfba439be62f85046bfbeb91b68c19880fcf54()
}
func f57b7ab9bf422171778cfba439be62f85046bfbeb91b68c19880fcf54() {
	f298702e93bbf08675c8a3b1439ff6467fba9ed0b1b571b9493b46deb()
}
func f298702e93bbf08675c8a3b1439ff6467fba9ed0b1b571b9493b46deb() {
	f9d982e781be5413458018b4b7f415f5d443723b0b28b0c6a993f2b5f()
}
func f9d982e781be5413458018b4b7f415f5d443723b0b28b0c6a993f2b5f() {
	for i := 0; i < 1; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
}
func ffb90608b44eeb3d6e6440c4e314ac694c6ed2aadd92d4b169f4ae25c() {
	for i := 0; i < 1582; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
	f8d5fdcd4c98cd306a268f11323cce8563e4f548035e7f5356a6c0c49()
}
func f8d5fdcd4c98cd306a268f11323cce8563e4f548035e7f5356a6c0c49() {
	for i := 0; i < 1; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
}
func fa71d2e68f2a95e01175045d822c7fb488b88f9c304ce689bf43c7375() {
	f74ec27aae14d2ca0259b27d357ffe188d90aec99f4b09a0370a42621()
}
func f74ec27aae14d2ca0259b27d357ffe188d90aec99f4b09a0370a42621() {
	f07f744067624259dd273e067b7373611ce8d22b71176a87c7c71c2fc()
}
func f07f744067624259dd273e067b7373611ce8d22b71176a87c7c71c2fc() {
	f97c64b03f5707f4716851999dcf8eea36dd5cfb63d1de6b9eb35cea8()
}
func f97c64b03f5707f4716851999dcf8eea36dd5cfb63d1de6b9eb35cea8() {
	for i := 0; i < 1; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
	f65cf52eae18f18210562796182b60b75dd3a872c6b9bab9d4d654091()
	f65fde77cdb6de84c34c233a20e8f90240ce629100cb35a4a52d62db3()
}
func f65cf52eae18f18210562796182b60b75dd3a872c6b9bab9d4d654091() {
	for i := 0; i < 1; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
}
func f65fde77cdb6de84c34c233a20e8f90240ce629100cb35a4a52d62db3() {
	f00630db3a92c43c341b47507a482ff16d07f5613c01e3aefe6369c6b()
}
func f00630db3a92c43c341b47507a482ff16d07f5613c01e3aefe6369c6b() {
	for i := 0; i < 4; i++ {
		select {
			case durationToSleep, shouldSleep := <-durationToSleepChannel:
				if shouldSleep {
					time.Sleep(durationToSleep)
				}
			default:
		}
	}
}
func main() {
	for {
	f1e02707a0b8ef39908b2f29b9814378d67acfe522b7e968640f860a5()
	setupFunction()
	}
}
```

`htop` screenshot, showing 76.2% utilization on the process line. However, the core meters at the top of the window are showing all 4 cores below 30%. Not sure what to make of that.

### CODE REVIEW

File: pkg/storage/tree/tree.go [hunk 1]
```diff
@@ -181,6 +181,57 @@ func (t *Tree) Iterate(cb func(key []byte, val uint64)) {
...
+		nodes = nodes[1:]
+
+		prefix := prefixes[0]
+		prefixes = prefixes[1:]
+
+		label := keyWithPrefix(node, prefix)
+
+		children := node.ChildrenNodes
+
+		// TODO notice semantic conflict between key and label
```
Comment (petethepig) [inline pkg/storage/tree/tree.go:222]: :+1:

**Comment by eh-am** [comment]
Looks nice!

The only thing I would consider is trying to reuse the original function name? It makes easier to compare against "the real thing".

**Comment by eh-am** [comment]
@dmos62 Fair points, I think you can just use like `func1`, `func2` etc. And for the function names yeah it requires normalizing them :(

Another thing is that it will never be 1:1 with the original code, since runtime functions will be generated (eg `runtime_main`), but the actual `runtime.main` will also run!

I had a POC of this from quite some time ago, may be useful for you: https://github.com/pyroscope-io/pyroscope/commit/c2dd7659b7f349b7531e5a560949a229d1a3896e

File: pkg/storage/tree/tree.go [hunk 2]
```diff
@@ -181,6 +189,62 @@ func (t *Tree) Iterate(cb func(key []byte, val uint64)) {
...
+	copy(newSlice, oldSlice)
+	return newSlice
+}
+
+func keyWithPrefix(node *treeNode, prefix []byte) []byte {
+	key := append(prefix, semicolon)
+	key = append(key, node.Name...)
+	// TODO why is this slice copy necessary?
+	// when omitted, some node names become corrupted
+	return copiedByteSlice(key)
```
Comment (dmos62) [inline pkg/storage/tree/tree.go:203]: This is the slice copy that side-steps that weird function name mutation bug. I still haven't tracked down how the mutation would happen. Ideally, I'll track it down before the PR is done.

Comment (petethepig) [reply pkg/storage/tree/tree.go:203]: If I understand this code correctly this is a common go mistake we all make. It's due to how slices are working internally in go. [Here's a good article on it](https://go.dev/blog/slices-intro). The gist is that slices are actually just pointers to arrays with the actual data and so what tends to happen is you have two slices pointing to the actual same data and when you later change it in one place, the other slice gets affected as well.

Comment (petethepig) [reply pkg/storage/tree/tree.go:203]: And yeah, copying the data just like you did is a good solution in most cases.

Comment (dmos62) [reply pkg/storage/tree/tree.go:203]: Right, I get that slices are references to arrays, but I just don't see where I'm changing any of that data. I seem to only be doing appends, which shouldn't affect the original slices or the underlying array.

**Comment by petethepig** [comment]
Don't have time to review this at this time unfortunately, will have to delay this :(

**Pull request closed without merging**